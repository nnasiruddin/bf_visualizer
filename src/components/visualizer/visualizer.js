import {ArrowKeyStepper, AutoSizer, Grid} from "react-virtualized";
import React from "react";

const Visualizer = props => {
    const {mode, scrollToColumn, scrollToRow, data} = props;

    const columnCount = data.length > 0 ? data.length : 10;

    const _getColumnWidth = () => {
        return 100;
    };

    const _getRowHeight = () => {
        return 30;
    };

    const _cellRenderer = ({
                               columnIndex,
                               key,
                               rowIndex,
                               scrollToColumn,
                               scrollToRow,
                               style,
                               data
                           }) => {

        const className = (rowIndex === scrollToRow && columnIndex === scrollToColumn) ? 'focused-cell' : 'cell';
        const content = (typeof data[columnIndex] !== 'undefined' || data[columnIndex] != null) ? data[columnIndex] : '';
        return (
            <span
                role="none"
                key={key}
                className={className}
                style={style}>

        {rowIndex === 0 ? columnIndex : content}

      </span>
        );
    };

    return (
        <ArrowKeyStepper
            columnCount={columnCount}
            rowCount={2}
            mode={mode}
            data={data}
            scrollToColumn={scrollToColumn}
            scrollToRow={scrollToRow}
        >
            {
                ({onSectionRendered, scrollToColumn, scrollToRow}) => (
                    <AutoSizer disableHeight>
                        {({width}) => (
                            <Grid
                                columnWidth={_getColumnWidth}
                                columnCount={columnCount}
                                height={62}
                                className='grid'
                                onSectionRendered={onSectionRendered}
                                cellRenderer={({columnIndex, key, rowIndex, style}) =>
                                    _cellRenderer({
                                        columnIndex,
                                        key,
                                        rowIndex,
                                        scrollToColumn,
                                        scrollToRow,
                                        style,
                                        data
                                    })
                                }
                                rowHeight={_getRowHeight}
                                rowCount={2}
                                scrollToColumn={scrollToColumn}
                                scrollToRow={scrollToRow}
                                width={width}
                            />
                        )}
                    </AutoSizer>
                )
            }
        </ArrowKeyStepper>
    );
};

export default Visualizer;
