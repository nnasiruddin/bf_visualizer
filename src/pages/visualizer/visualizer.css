.grid {
    border: 1px solid #e0e0e0;
}

.cell {
    box-sizing: content-box;
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    border: none;
    border-right: 1px solid #e0e0e0;
    border-bottom: 1px solid #e0e0e0;
}

.focused-cell {
    background-color: #e0e0e0;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    border: none;
    border-right: 1px solid #e0e0e0;
    border-bottom: 1px solid #e0e0e0;
    font-weight: bold;
}

.visualize-main {
    width: 1000px;
    padding-left: 10%;
    padding-top: 5%;
}

.visualize-header {
    text-align: center;
}

.visualize-footer {
    margin-top: 2%;
}

.visualize-form {
    margin-top: 2%;
}

.visualize-display {
    margin-top: 2%;
}

textarea {
    min-height: 300px;
}

.visualize-output {
    margin-top: 2%;
    min-height: 100px;
    border: 1px solid #e0e0e0;
    border-radius: 5px;
}

.visualize-error {
    text-align: center;
    margin-top: 10%;
}

.warning-sign {
    font-size: 50px;
    color: red;
}
