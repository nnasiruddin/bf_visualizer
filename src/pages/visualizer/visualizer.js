import React from 'react';
import './visualizer.css';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import 'react-virtualized/styles.css'; // only needs to be imported once
import Visualizer from '../../components/visualizer';
import { Button, FormControl, FormGroup, ControlLabel, Grid, Row, Col, Glyphicon } from 'react-bootstrap';
import * as visualizerActions from '../../actions/visualizerActions';

export const VisualizerPage = props => {
    const { data, scrollToRow, scrollToColumnScript, scrollToColumnData, mode , editMode, script, output, done, showError} = props;

    const handleScriptInput = (value) => {
        props.actions.updateInputScript(value);
    };

    const handleExecute = () => {
      props.actions.loadBrainFuck()
    };

    const handleStep = () => {
        props.actions.step()
    };

    const handlePlay = () => {
        props.actions.play()
    };

    const handleStop = () => {
      props.actions.stop()
    };

    const generateForm = () => {
        return (
            <div>
                <div className='visualize-form'>
                    <Grid>
                        <Row>
                            <Col xs={10}>
                                <form>
                                    <FormGroup controlId="visualizeScript">
                                        <ControlLabel>Script</ControlLabel>
                                        <FormControl componentClass="textarea" onChange={(event) => handleScriptInput(event.target.value)} placeholder="Enter Script Here..." />
                                    </FormGroup>
                                </form>
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <div className='visualize-footer'>
                                    <Button onClick={() => handleExecute()}>Execute Script</Button>
                                </div>
                            </Col>
                        </Row>
                    </Grid>
                </div>
            </div>
        )
    };

    const generateDisplay = (data, scrollToRow, scrollToColumn, mode, output, done) => {
        return (
            <div className='visualize-display'>
                <Grid>
                    <Row>
                        <Col xs={4}>
                            <h3>Script:</h3>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={10}>
                            <Visualizer data={data} scrollToRow={scrollToRow} scrollToColumn={scrollToColumn} mode={mode} />
                        </Col>
                    </Row>
                    <Row className='visualize-display'>
                        { !done &&
                            <Col xs={3}>
                                <Button onClick={() => handleStep()}>Step</Button>
                            </Col>
                        }
                        {
                            !done &&
                            <Col xs={3}>
                                <Button onClick={() => handlePlay()}>Play</Button>
                            </Col>
                        }
                    </Row>
                    <Row>
                        <Col xs={10}>
                            <div>
                                <h3>Output:</h3>
                            </div>
                            <div className='visualize-output'>
                                <div>
                                    {output}
                                </div>
                            </div>
                        </Col>
                    </Row>
                </Grid>
            </div>
        )
    };

    const generateError = () => {
      return (
          <div className='visualize-error'>
              <Glyphicon style={{fontSize: '80px', color: 'red'}} glyph="exclamation-sign" />
              <h3>Oops! Its not you its us, that code brainfucked our compiler, please reload the page to restart</h3>
          </div>
      )
    };

    const generateMainScreen = () => {
      return (
          <div>
              <div className='visualize-header'>
                  <h2>Brain Fuck Visualizer</h2>
              </div>
              <Grid>
                  <Row>
                      <Col xs={4}>
                          <h3>Data:</h3>
                      </Col>
                  </Row>
                  <Row>
                      <Col xs={10}>
                          <Visualizer data={data} scrollToRow={scrollToRow} scrollToColumn={scrollToColumnData} mode={mode} />
                      </Col>
                  </Row>
                  <Row>
                      <Col xs={10}>

                      </Col>
                  </Row>
              </Grid>
              {
                  editMode && generateForm()
              }
              {
                  !editMode && generateDisplay(script, scrollToRow, scrollToColumnScript, mode, output, done)
              }
          </div>
      )
    };


    return (
        <div className='visualize-main'>
            {done && handleStop()}
            {showError && generateError()}
            {!showError && generateMainScreen()}
        </div>
    )
};

function mapStateToProps(state) {
    return {
        data: state.visualizeReducer.data,
        scrollToColumnScript: state.visualizeReducer.scrollToColumnScript,
        scrollToColumnData: state.visualizeReducer.scrollToColumnData,
        scrollToRow: state.visualizeReducer.scrollToRow,
        editMode: state.visualizeReducer.editMode,
        script: state.visualizeReducer.script,
        inputScript: state.visualizeReducer.inputScript,
        output: state.visualizeReducer.output,
        done: state.visualizeReducer.done,
        showError: state.visualizeReducer.showError
    }
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            ...visualizerActions
        }, dispatch)
    }
}

VisualizerPage.propTypes = {
    data: PropTypes.array.isRequired,
    scrollToColumnScript: PropTypes.number.isRequired,
    scrollToColumnData: PropTypes.number.isRequired,
    scrollToRow: PropTypes.number.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(VisualizerPage);
