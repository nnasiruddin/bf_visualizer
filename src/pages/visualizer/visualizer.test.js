import React from 'react';
import renderer from 'react-test-renderer';
import initState from '../../reducers/initialState';
import { VisualizerPage } from './visualizer';

describe('Visualizer rendering test', () => {

    it('should render Dataset Page component correctly', () => {
        const tree = renderer
            .create(
                <VisualizerPage
                    mode={initState.visualizer.mode}
                    scrollToRow={initState.visualizer.scrollToRow}
                    scrollToColumnData={initState.visualizer.scrollToColumnData}
                    scrollToColumnScript={initState.visualizer.scrollToColumnScript}
                    editMode={initState.visualizer.editMode}
                    inputScript={initState.visualizer.inputScript}
                    output={initState.visualizer.output}
                    id={initState.visualizer.id}
                    data={initState.visualizer.data}
                    script={initState.visualizer.script}
                >
                </VisualizerPage>
            ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

