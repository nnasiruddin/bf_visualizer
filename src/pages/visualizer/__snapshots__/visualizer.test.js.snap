// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizer rendering test should render Dataset Page component correctly 1`] = `
<div
  className="visualize-main"
>
  <div>
    <div
      className="visualize-header"
    >
      <h2>
        Brain Fuck Visualizer
      </h2>
    </div>
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-4"
        >
          <h3>
            Data:
          </h3>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-10"
        >
          <div
            onKeyDown={[Function]}
          >
            <div
              style={
                Object {
                  "overflow": "visible",
                  "width": 0,
                }
              }
            >
              <div
                aria-label="grid"
                aria-readonly={true}
                className="ReactVirtualized__Grid grid"
                onScroll={[Function]}
                role="grid"
                style={
                  Object {
                    "WebkitOverflowScrolling": "touch",
                    "boxSizing": "border-box",
                    "direction": "ltr",
                    "height": 62,
                    "overflowX": "auto",
                    "overflowY": "hidden",
                    "position": "relative",
                    "width": 0,
                    "willChange": "transform",
                  }
                }
                tabIndex={0}
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-10"
        />
      </div>
    </div>
    <div>
      <div
        className="visualize-form"
      >
        <div
          className="container"
        >
          <div
            className="row"
          >
            <div
              className="col-xs-10"
            >
              <form>
                <div
                  className="form-group"
                >
                  <label
                    className="control-label"
                    htmlFor="visualizeScript"
                  >
                    Script
                  </label>
                  <textarea
                    className="form-control"
                    id="visualizeScript"
                    onChange={[Function]}
                    placeholder="Enter Script Here..."
                  />
                </div>
              </form>
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="col-xs-3"
            >
              <div
                className="visualize-footer"
              >
                <button
                  className="btn btn-default"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Execute Script
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
